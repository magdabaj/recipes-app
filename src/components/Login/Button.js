import React from 'react';
import styled from 'styled-components';

const Button = styled.button`
  background-color: #9a1750;
  color: white;
  border: 1px solid #9a1750;
  text-transform: uppercase;
  font-size: 14px;
  font-weight: 600;
  padding: 10px 15px;
  cursor: pointer;
  margin: 20px;
  border-radius: 5px;

  &:hover {
    background-color: white;
    color: #9a1750;
  }
  
  @media (min-width: 768px) {
    width: 100%;
    display: flex;
    justify-content: center;
  }
`;

export const ButtonSmall = styled(Button)`
  padding: 5px 8px;

  @media (min-width: 768px) {
    font-size: 10px;
    margin-left: 5px;
    margin: 0;
    width: 50%
  }
`;

export default Button;
