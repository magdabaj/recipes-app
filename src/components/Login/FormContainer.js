// eslint-disable-next-line
import React from 'react';
import styled from 'styled-components';

const FormContainer = styled.div`
  display: flex;
  align-items: center;
  // flex-direction: row;
  // justify-content: center;
  // align-content: space-evenly;
  padding: 20px;

  @media (max-width: 768px) {
    width: 100%;
    padding: 0;
  }
`;

export default FormContainer;
