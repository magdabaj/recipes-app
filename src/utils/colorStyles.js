const colorStyles = {
    darkRed: '#5D001E',
    lightGrey: '#E3E2DF',
    lightPink: '#E3AFBC',
    mediumPink: '#EE4C7C',
    darkPink: '#9A1750',
};

export default colorStyles;
