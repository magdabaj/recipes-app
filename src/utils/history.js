// import { createBrowserHistory } from 'history';
//
// const history = createBrowserHistory();

import { createMemoryHistory } from "history";

const history = createMemoryHistory()

export default history;
