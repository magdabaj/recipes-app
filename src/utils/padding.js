const basePadding = 8;

const padding = {
    extraSmall: `${basePadding / 4}px`, // 2px
    small: `${basePadding / 2}px`, // 4px
    base: `${basePadding}px`, // 8px
    medium: `${basePadding * 1.5}px`, // 12px
    large: `${basePadding * 2}px`, // 16px
}

export default padding;