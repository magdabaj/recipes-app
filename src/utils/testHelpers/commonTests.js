const commonTests = (render, useFirstChildAsSnapshot = true) => {
    it('does not log errors in console', () => {
        const spy = jest.spyOn(global.console, 'error')
        render()
        expect(spy).not.toHaveBeenCalled()
    })

    it('should render and match the snapshot',  () => {
        const { container } = render()
        expect(container).not.toBeNull()
        if(useFirstChildAsSnapshot) expect(container.firstChild).not.toBeNull()
        expect(
            useFirstChildAsSnapshot ? container.firstChild : container,
        ).toMatchSnapshot()
    });
}

export default commonTests;