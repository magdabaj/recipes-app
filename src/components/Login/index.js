/**
 *
 * LoginFormComponent
 *
 */

import React, { useState } from 'react';
import TextInput from '../TextInput';
import '../../containers/App/index.css';
import Button, { ButtonSmall } from './Button';
import fetchStates from '../../utils/fetchStates'
import {Redirect} from "react-router";
import {toast} from "react-toastify";

function Login({ user, signIn, signUp, error, setError, status }) {
    const [userSubmit, setUser] = useState({
        email: '',
        password: '',
    });
    const [errors, setErrors] = useState({});

    const handleChange = event => {
        const { name, value } = event.target;

        setUser(prevUser => ({
            ...prevUser,
            [name]: value,
        }));
    };

    const formIsValid = () => {
        const { email, password } = userSubmit;
        const errors = {};

        if (!email) errors.email = 'Email is required';
        if (!password) errors.password = 'Password is required';

        setErrors(errors);

        return Object.keys(errors).length === 0;
    };

    function handleSave(event) {
        event.preventDefault();
        if (!formIsValid()) return;
        try {
            signUp({ email: userSubmit.email, password: userSubmit.password });
        } catch (error) {
            setErrors({ handleSave: error.message });
        }
    }

    function handleSignIn(event) {
        event.preventDefault();
        if (!formIsValid()) return;
        try {
            signIn({ email: userSubmit.email, password: userSubmit.password });
        } catch (error) {
            setErrors({ handleSave: error.message });
        }
    }

    // if (error) {
    //   toast.error(error);
    //   setError();
    // }
    return (
        <div className={'form-container'}>
            <form className={'form'} onSubmit={handleSignIn}>
                <TextInput
                    type={'text'}
                    name={'email'}
                    label={'Email'}
                    value={userSubmit.email}
                    onChange={handleChange}
                    error={errors.email}
                    placeholder={'Email'}
                />
                <TextInput
                    type={'password'}
                    name={'password'}
                    label={'Hasło'}
                    value={userSubmit.password}
                    onChange={handleChange}
                    error={errors.password}
                    placeholder={"Hasło"}
                />
                <Button type={'submit'} onSubmit={handleSignIn}>
                    {status === fetchStates.fetching
                        ? 'Ladowanie...'
                        : 'Zaloguj się'}
                </Button>
                <div className={'sign-up'}>
                    <div>Nie masz jeszcze konta?</div>
                    <ButtonSmall type={'button'} onClick={handleSave}>
                        {status === fetchStates.fetching
                            ? 'Ladowanie...'
                            : 'Zarejestruj się'}
                    </ButtonSmall>
                </div>
            </form>
            {status === fetchStates.success && <Redirect to={'/user-recipes'}/>}
            {/*{status === fetchStates.error && toast.error(error)}*/}
        </div>
    );
}

Login.propTypes = {};

export default Login;
