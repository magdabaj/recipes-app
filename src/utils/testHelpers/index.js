import { Router } from 'react-router-dom'
import { createMemoryHistory } from "history";
import store from "../store";
import {Provider} from "react-redux";
import React from "react";
import { render } from "@testing-library/react";
import {createStore} from "redux";
// import {reducer} from "react-toastify/dist/hooks/toastContainerReducer";

// const store  = createStore(rootReducer, initialState={})
const memoryHistory = createMemoryHistory({
    initialEntries: ['/'],
})

export const withStoreAndRouter = component =>
    <Router history={memoryHistory}>
        <Provider store={store}>
            {component}
        </Provider>
    </Router>

export const renderWithRouter = (
    ui,
    {
        history = memoryHistory,
    } = {}
) => (
    {
        ...render(<Router history={history}>{ui}</Router>),
        history,
    }
)

export const renderWithRouterKentCDodds = (
    ui,
        {
            route='/',
            history = createMemoryHistory({initialEntries: [route]}),
            ...renderOptions
        } = {}
) => {
    // eslint-disable-next-line react/prop-types
    function Wrapper({children}) {
        return <Router history={history}>
            {children}
        </Router>
    }

    return {
        ...render(ui, {wrapper: Wrapper, ...renderOptions})
    }
}

export const renderWithProviderKentCDodds = (
        ui,
        reducer,
        {initialState, store = createStore(reducer, initialState), ...options} = {},
    ) => {
    // eslint-disable-next-line react/prop-types
    function Wrapper({children}) {
        return <Provider store={store}>
            {children}
        </Provider>
    }
    return render(ui, {wrapper: Wrapper, ...options})
}