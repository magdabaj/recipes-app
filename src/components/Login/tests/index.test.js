import React from "react";
import { render, screen } from "@testing-library/react";
import Login from "../index";
import * as jestDOM from '@testing-library/jest-dom';
import fetchStates from "../../../utils/fetchStates";
import { renderWithRouter } from "../../../utils/testHelpers";
import commonTests from "../../../utils/testHelpers/commonTests";
import {axe} from "jest-axe";
import userEvent from "@testing-library/user-event";

expect.extend(jestDOM)

const signInMocked = jest.fn()
const signUpMocked = jest.fn()


jest.mock('axios')

const renderLogin = () =>
    render(<Login signIn={signInMocked} signUp={signUpMocked}/>)

commonTests(renderLogin)

test('checks if form has any violations', async () => {
    const { container } = renderLogin()
    const results = await axe(container)
    expect(results).toHaveNoViolations()
})

test('renders Login form', () => {
    render(<Login signIn={signInMocked} signUp={signUpMocked}/>)
    const loginButton = screen.getByText(/Zaloguj się/i)
    const registerButton = screen.getByText(/Zarejestruj się/i)

    expect(loginButton).toBeInTheDocument()
    expect(registerButton).toBeInTheDocument()
})

test('renders Login form', () => {
    render(<Login signIn={signInMocked} signUp={signUpMocked} status={fetchStates.fetching}/>)
    const buttons = screen.getAllByText(/ladowanie.../i)

    expect(buttons[1]).toBeInTheDocument()
    expect(buttons[0]).toBeInTheDocument()
})

test('redirects if user logs in successfully', async () => {
    const redirectUrl = '/user-recipes'

    const { history } = renderWithRouter(<Login signIn={signInMocked} signUp={signUpMocked} status={fetchStates.success}/> )

    expect(history.location.pathname).toEqual(redirectUrl)
})

test('sends signUp action if form is valid', () => {
    renderLogin()
    const emailInput = screen.getByLabelText(/email/i)
    const passwordInput = screen.getByLabelText(/hasło/i)
    const signInButton = screen.getByText(/zaloguj się/i)

    userEvent.type(emailInput, 'test')
    userEvent.type(passwordInput, 'test')
    userEvent.click(signInButton)
    expect(signInMocked).toHaveBeenCalled()
    expect(signInMocked).toHaveBeenCalledTimes(1)
    expect(signInMocked).toHaveBeenCalledWith({
        email: 'test',
        password: 'test',
    })
})

test('shows error message when trying to submit empty form', () => {
    renderLogin()
    const signInButton = screen.getByText(/zaloguj się/i)
    let errorMessage = screen.queryByText(/musisz wpisać email/i)

    expect(errorMessage).not.toBeInTheDocument()
    userEvent.click(signInButton)
    expect(signInMocked).not.toHaveBeenCalled()
    errorMessage = screen.getByText(/musisz wpisać email/i)
    expect(errorMessage).toBeInTheDocument()
})


// todo should I test toasts?