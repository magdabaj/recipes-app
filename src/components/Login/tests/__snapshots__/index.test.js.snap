// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render and match the snapshot 1`] = `
<div
  class="form-container"
>
  <form
    class="form"
  >
    <div
      class="input-data"
    >
      <label
        class="login-label"
        for="email"
      >
        Email
      </label>
      <div
        class="field"
      >
        <input
          class="input"
          id="email"
          name="email"
          placeholder="Email"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="input-data"
    >
      <label
        class="login-label"
        for="password"
      >
        Hasło
      </label>
      <div
        class="field"
      >
        <input
          class="input"
          id="password"
          name="password"
          placeholder="Hasło"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      class="Button-sc-9ry37c-0 hcJNuE"
      type="submit"
    >
      Zaloguj się
    </button>
    <div
      class="sign-up"
    >
      <div>
        Nie masz jeszcze konta?
      </div>
      <button
        class="Button-sc-9ry37c-0 Button__ButtonSmall-sc-9ry37c-1 gunhDf"
        type="button"
      >
        Zarejestruj się
      </button>
    </div>
  </form>
</div>
`;
